stream_alert/classifier/parsers.py

```diff
@@ -0,0 +1,807 @@
+
+    @property
+    def json_path(self):
+        return self.configuration.get('json_path')
+
+    @property
+    def envelope_schema(self):
+        return self.configuration.get('envelope_keys', {})
+
+    @property
+    def optional_envelope_keys(self):
+        return set(self.configuration.get('optional_envelope_keys', {}))
+
+    @property
+    def valid(self):
+        return len(self._valid_parses) != 0
+
+    @property
+    def invalid_parses(self):
+        return self._invalid_parses
+
+    @property
+    def parses(self):
+        return self._valid_parses
+
+    @classmethod
+    def default_optional_values(cls, key):
+        """Return a default value for a given supported schema type"""
+        # Return instances of types that are defined, or of the type being passed in
+        return cls._TYPE_MAP[key]() if isinstance(key, basestring) else type(key)()
+
+    @classmethod
+    def _apply_envelope(cls, record, envelope):
+        """Apply the envelope extracted from the orignal record to the parsed record
+
+        Args:
+            record (dict): Value of the successfully parsed record
+            envelope (dict): Extracted envelope to be added to the parsed record
+        """
+        if not envelope:
+            return
+
+        record.update({cls.ENVELOPE_KEY: envelope})
+
+    @classmethod
+    def _add_optional_keys(cls, data, schema, optional_keys):
+        """Add optional keys to a raw record
+
+        Args:
+            data (dict): Parsed record value
+            schema (dict): Schema for the data being parsed
+            optional_keys (set): The optional keys in the schema
+        """
+        if not (schema and optional_keys):
+            return  # Nothing to do
+
+        for key_name in optional_keys:
+            # Add the optional key if it does not exist in the parsed json payload
+            data[key_name] = data.get(key_name, cls.default_optional_values(schema[key_name]))
+
+    @classmethod
+    def _matches_log_patterns(cls, record, log_patterns):
+        """Check if all log patterns specified for this record match
+
+        Args:
+            record (dict): Parsed record value
+            log_patterns (dict): Log patterns that should be enforced for the data
+            being parsed. This could be checking a nested value, so it should be
+            passed in as an argument.
+
+        Returns:
+            bool: True if all specified log patterns match, False otherwise
+        """
+        # Return True immediately if there are no log patterns
+        if not log_patterns:
+            return True
+
+        result = True
+        for field, patterns in log_patterns.iteritems():
+            # handle nested log_patterns
+            if isinstance(patterns, dict):
+                return cls._matches_log_patterns(record[field], patterns)
+
+            # Transform a flat pattern into a list
+            if isinstance(patterns, basestring):
```

Comment:
hmm good catch. I'm hesitant to make the change, since I like the simplicity, and since the `2to3` conversion utility that we'll use when we 'upgrade' will actually handle the conversion for us